File: Parallel-Csharp/src/Parallel/Linq101/Orderings.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101;

public class Orderings
{
    public List<Product> GetProductList() => Products.ProductList;

    public List<Customer> GetCustomerList() => Customers.CustomerList;

    public int OrderbySyntax()
    {
        string[] words = { "cherry", "apple", "blueberry" };

        var sortedWords = words.OrderBy(w => w).ToArray();

        Console.WriteLine("The sorted list of words:");
        foreach (var w in sortedWords)

            Console.WriteLine(w);
        return 0;
    }

    public int OrderbyProperty()
    {
        string[] words = { "cherry", "apple", "blueberry" };

        var sortedWords = words.OrderBy(w => w.Length).ToArray();

        Console.WriteLine("The sorted list of words (by length):");
        foreach (var w in sortedWords)
            Console.WriteLine(w);

        return 0;
    }

    public int OrderByProducts()
    {
        List<Product> products = GetProductList();
        var sortedProducts = products.OrderBy(p => p.ProductName).ToList();
        foreach (var product in sortedProducts)
            Console.WriteLine(product);
        return 0;
    }

    public class CaseInsensitiveComparer : IComparer<string>
    {
        public int Compare(string x, string y) =>
            string.Compare(x, y, StringComparison.OrdinalIgnoreCase);
    }

    public int OrderByWithCustomComparer()
    {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };

        var sortedWords = words.OrderBy(a => a, new CaseInsensitiveComparer()).ToArray();

        foreach (var word in sortedWords)
            Console.WriteLine(word);

        return 0;
    }

    public int OrderByDescendingSyntax()
    {
        double[] doubles = { 1.7, 2.3, 1.9, 4.1, 2.9 };

        var sortedDoubles = doubles.OrderByDescending(d => d).ToArray();

        Console.WriteLine("The doubles from highest to lowest:");

        foreach (var d in sortedDoubles)
            Console.WriteLine(d);

        return 0;
    }

    public int OrderProductsDescending()
    {
        List<Product> products = GetProductList();

        var sortedProducts = products.OrderByDescending(p => p.UnitsInStock).ToArray();

        foreach (var product in sortedProducts)
            Console.WriteLine(product);

        return 0;
    }

    public int DescendingCustomComparer()
    {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };

        var sortedWords = words.OrderByDescending(a => a, new CaseInsensitiveComparer()).ToArray();

        foreach (var word in sortedWords)
            Console.WriteLine(word);

        return 0;
    }

    public int ThenBySyntax()
    {
        string[] digits = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

        var sortedDigits = digits.OrderBy(d => d.Length).ThenBy(d => d).ToArray();

        Console.WriteLine("Sorted digits:");
        foreach (var d in sortedDigits)
            Console.WriteLine(d);

        return 0;
    }

    public int ThenByCustom()
    {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };

        var sortedWords = words
            .OrderBy(a => a.Length).ThenBy(a => a, new CaseInsensitiveComparer()).ToArray();

        foreach (var word in sortedWords)
            Console.WriteLine(word);

        return 0;
    }

    public int ThenByDifferentOrdering()
    {
        List<Product> products = GetProductList();

        var sortedProducts = products.OrderBy(p => p.Category).ThenByDescending(p => p.UnitPrice).ToList();

        foreach (var product in sortedProducts)
            Console.WriteLine(product);

        return 0;
    }

    public int CustomThenByDescending()
    {
        string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };

        var sortedWords = words.OrderBy(a => a.Length)
            .ThenByDescending(a => a, new CaseInsensitiveComparer()).ToArray();

        foreach (var word in sortedWords)
            Console.WriteLine(word);

        return 0;
    }

    public int OrderingReversal()
    {
        string[] digits = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

        var reversedIDigits = digits.Where(d => d[1] == 'i').Reverse().ToArray();

        Console.WriteLine("A backwards list of the digits with a second character of 'i':");
        foreach (var d in reversedIDigits)
            Console.WriteLine(d);

        return 0;
    }
}

File: Parallel-Java/src/parallel/linq101/Orderings.java
package parallel.linq101;

import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class Orderings {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public int orderbySyntax() {
        String[] words = {"cherry", "apple", "blueberry"};
        var sortedWords = Arrays.stream(words).sorted().toArray();
        System.out.println("The sorted list of words:");
        for (var w : sortedWords)
            System.out.println(w);
        return 0;
    }

    public int orderbyProperty() {
        String[] words = {"cherry", "apple", "blueberry"};
        var sortedWords = Arrays.stream(words).sorted(Comparator.comparing(w -> w.length())).toArray();
        System.out.println("The sorted list of words (by length):");
        for (var w : sortedWords)
            System.out.println(w);
        return 0;
    }

    public int orderByProducts() {
        List<Product> products = getProductList();
        var sortedProducts = products.stream().sorted(Comparator.comparing(p -> p.getProductName()))
                .collect(Collectors.toList());
        for (var product : sortedProducts)
            System.out.println(product);
        return 0;
    }

    public static class CaseInsensitiveComparer implements Comparator<String> {
        @Override
        public int compare(String x, String y) {
            return x.toUpperCase().compareTo(y.toUpperCase());
        }
    }

    public int orderByWithCustomComparer() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        var sortedWords = Arrays.stream(words).sorted(new CaseInsensitiveComparer()).toArray();
        for (var word : sortedWords)
            System.out.println(word);
        return 0;
    }

    public int orderByDescendingSyntax() {
        double[] doubles = {1.7, 2.3, 1.9, 4.1, 2.9};
        var sortedDoubles = Arrays.stream(doubles).boxed().sorted(Comparator.reverseOrder()).toArray();
        System.out.println("The doubles from highest to lowest:");
        for (var d : sortedDoubles)
            System.out.println(d);
        return 0;
    }

    public int orderProductsDescending() {
        List<Product> products = getProductList();
        var sortedProducts = products.stream().sorted(
                Comparator.comparing((Product p) -> p.getUnitsInStock(), Comparator.reverseOrder())).toList();
        for (var product : sortedProducts)
            System.out.println(product);
        return 0;
    }

    public int descendingCustomComparer() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        var sortedWords = Arrays.stream(words).sorted(new CaseInsensitiveComparer().reversed()).toArray();
        for (var word : sortedWords)
            System.out.println(word);
        return 0;
    }

    public int thenBySyntax() {
        String[] digits = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        var sortedDigits = Arrays.stream(digits).sorted(Comparator.comparing((String w) -> w.length())
                .thenComparing(d -> d)).toArray();
        System.out.println("Sorted digits:");
        for (var d : sortedDigits)
            System.out.println(d);
        return 0;
    }

    public int thenByCustom() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        var sortedWords = Arrays.stream(words).sorted(Comparator.comparing((String w) -> w.length())
                .thenComparing(new CaseInsensitiveComparer())).toArray();
        for (var word : sortedWords)
            System.out.println(word);
        return 0;
    }

    public int thenByDifferentOrdering() {
        List<Product> products = getProductList();
        var sortedProducts = products.stream().sorted(Comparator.comparing((Product p) -> p.getCategory())
                .thenComparing(p -> p.getUnitPrice(), Comparator.reverseOrder())).collect(Collectors.toList());
        for (var product : sortedProducts)
            System.out.println(product);
        return 0;
    }

    public int customThenByDescending() {
        String[] words = {"aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry"};
        var sortedWords = Arrays.stream(words).sorted(Comparator.comparing((String w) -> w.length()).
                thenComparing(new CaseInsensitiveComparer().reversed())).toArray();
        for (var word : sortedWords)
            System.out.println(word);
        return 0;
    }

    public int orderingReversal() {
        String[] digits = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        var reversedIDigits = Arrays.stream(digits).filter(d -> d.charAt(1) == 'i').
                collect(Collectors.collectingAndThen(Collectors.toList(),
                        list->{Collections.reverse(list);return list.stream();})).toArray();
        System.out.println("A backwards list of the digits with a second character of 'i':");
        for (var d : reversedIDigits)
            System.out.println(d);
        return 0;
    }
}
